File: examples/authentication/cc/rt/naeem/hottentot/runtime/service/service_runtime.cc
#include <iostream>

#include "service_runtime.h"
#include "service.h"
#include "tcp_server.h"
#include "default_tcp_server_factory.h"


namespace naeem {
  namespace hottentot {
    namespace runtime {
      namespace service {
        TcpServerFactory* ServiceRuntime::tcpServerFactory_ = 0;
        std::map<Endpoint, std::vector<Service*>*, Endpoint::Comparator> ServiceRuntime::services_;
        std::map<Endpoint, std::map<uint8_t, RequestHandler*>*, Endpoint::Comparator> requestHandlers_;
        void
        ServiceRuntime::Init(int argc,
                             char **argv) {
        }
        void
        ServiceRuntime::Register(std::string   host, 
                                 uint32_t      port, 
                                 Service       *service) {
          Endpoint endpoint(host, port);
          if (services_.count(endpoint) == 0) {
            std::vector<Service*> *e = new std::vector<Service*>();
            e->push_back(service);
            services_.insert(std::pair<Endpoint, std::vector<Service*>*>(endpoint, e));
          } else {
            services_.find(endpoint)->second->push_back(service);
          }
        }
        void
        ServiceRuntime::Start() {
          for (std::map<Endpoint, std::vector<Service*>*, Endpoint::Comparator>::iterator it = services_.begin();
               it != services_.end();
               it++) {
            TcpServer *tcpServer = GetTcpServerFactory()->CreateTcpServer(it->first.GetHost(), 
                                                                          it->first.GetPort(), 
                                                                          it->second);
            tcpServer->BindAndStart();
            std::cout << "Endpoint started: " << it->first.GetHost() << ":" << it->first.GetPort() << std::endl;
          }
          while(true);
        }
        TcpServerFactory*
        ServiceRuntime::GetTcpServerFactory() {
          if (tcpServerFactory_ == 0) {
            tcpServerFactory_ = new DefaultTcpServerFactory;
          }
          return tcpServerFactory_;
        }
      }
    }
  }
}

File: examples/authentication/java/src/main/java/runtime/service/ServiceRuntime.java
package runtime.service;


import runtime.Endpoint;
import runtime.ProtocolFactory;
import runtime.RequestHandler;

import java.io.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ServiceRuntime {

    private Map<Endpoint,List<Service>> serviceMap = new HashMap<Endpoint, List<Service>>();
    private Map<Endpoint , Map<Integer , RequestHandler>> requestHandlerMap = new HashMap<Endpoint, Map<Integer, RequestHandler>>();
    public void register(String host , int port , Service service){
        //TODO

        Endpoint endPoint = new Endpoint();
        endPoint.setHost(host);
        endPoint.setPort(port);


        if(serviceMap.containsKey(endPoint)){
            List<Service> serviceListInEndpoint = serviceMap.get(endPoint);
            //TODO search in service list for duplicate service registered
            serviceListInEndpoint.add(service);
            Map<Integer, RequestHandler> requestHandlerMapInEndpoint = requestHandlerMap.get(endPoint);
            requestHandlerMapInEndpoint.put(service.getServiceId() , service.getRequestHandler());
        }else{
            ArrayList<Service> serviceList = new ArrayList<Service>();
            serviceList.add(service);
            serviceMap.put(endPoint,serviceList);
            HashMap<Integer, RequestHandler> requestHandlers = new HashMap<Integer , RequestHandler>();
            requestHandlers.put(service.getServiceId() , service.getRequestHandler());
            requestHandlerMap.put(endPoint, requestHandlers);
        }

    }
    public void start() throws IOException {
        //TODO  ????!!!!

        for(Map.Entry<Endpoint , Map<Integer , RequestHandler> > entry : requestHandlerMap.entrySet()){
            Map<Integer, RequestHandler> requestHandlerMapInEndpoint = entry.getValue();
            Endpoint endpoint = entry.getKey();
            TcpServer tcpServer = TcpServerFactory.create(endpoint.getHost() , endpoint.getPort() , requestHandlerMapInEndpoint );
            tcpServer.bindAndStart();
            tcpServer.setProtocol(ProtocolFactory.create());
        }
//        for (ServiceInfo serviceInfo : servicePool){
//            TcpServer tcpServer = TcpServerFactory.create(serviceInfo);
//            tcpServer.bindAndStart();
//            tcpServer.setProtocol(ProtocolFactory.create());
//        }
    }
}
